onadata/libs/utils/csv_import.py

```diff
@@ -168,160 +169,50 @@ def submit_csv(username, xform, csv_file, overwrite=False):
     :return: If sucessful, a dict with import summary else dict with error str.
     :rtype: Dict
     """
-    if isinstance(csv_file, str):
-        csv_file = BytesIO(csv_file)
-    elif csv_file is None or not hasattr(csv_file, 'read'):
-        return async_status(FAILED, (u'Invalid param type for `csv_file`. '
-                                     'Expected utf-8 encoded file or unicode'
-                                     ' string got {} instead.'
-                                     .format(type(csv_file).__name__)))
-
-    num_rows = sum(1 for row in csv_file) - 1
-    csv_file.seek(0)
-
-    csv_reader = ucsv.DictReader(csv_file, encoding='utf-8-sig')
-    csv_header = csv_reader.fieldnames
-
-    # check for spaces in headers
-    if any(' ' in header for header in csv_header):
-        return async_status(FAILED,
-                            u'CSV file fieldnames should not contain spaces')
-
-    # Get the data dictionary
-    xform_header = xform.get_headers()
-
-    missing_col = set(xform_header).difference(csv_header)
-    addition_col = set(csv_header).difference(xform_header)
-
-    # change to list
-    missing_col = list(missing_col)
-    addition_col = list(addition_col)
-    # remove all metadata columns
-    missing = [
-        col for col in missing_col
-        if not col.startswith("_") and col not in IGNORED_COLUMNS
-    ]
-
-    # remove all metadata inside groups
-    missing = [col for col in missing if '/_' not in col]
-
-    # ignore if is multiple select question
-    for col in csv_header:
-        # this col is a multiple select question
-        survey_element = xform.get_survey_element(col)
-        if survey_element and \
-                survey_element.get('type') == MULTIPLE_SELECT_TYPE:
-            # remove from the missing and additional list
-            missing = [x for x in missing if not x.startswith(col)]
-
-            addition_col.remove(col)
-
-    # remove headers for repeats that might be missing from csv
-    missing = sorted([m for m in missing if m.find('[') == -1])
-
-    # Include additional repeats
-    addition_col = [a for a in addition_col if a.find('[') == -1]
+    # Validate csv before creation of Instances
+    try:
+        validated_data = validate_csv(csv_file, xform)
+    except UnicodeDecodeError:
+        return async_status(
+            FAILED, 'CSV file must be utf-8 encoded')
 
-    if missing:
+    if not validated_data.get('valid'):
         return async_status(
-            FAILED, u"Sorry uploaded file does not match the form. "
-            u"The file is missing the column(s): "
-            u"{0}.".format(', '.join(missing)))
+            FAILED,
+            validated_data.get('error_msg')
+        )
 
     if overwrite:
         xform.instances.filter(deleted_at__isnull=True)\
             .update(deleted_at=timezone.now(),
                     deleted_by=User.objects.get(username=username))
 
+    validated_rows = validated_data.get('data')
```

Comment:
Are we having all these rows in memory? I see likely hood of memory heavy implementation, perhaps reading from a buffer or reading through the file again once we know all the records are valid could ensure we have a small memory footprint.

Reply:
There are a few things we do to the data during the validation process like making sure the `date` and `dateTime` datatypes are `isoformatted` not quite sure how we can do that with out in someway storing the data in memory.

Reply:
I believe @ukanga's point is to limit the amount of data in memory at any one point in time, not to keep it all out of memory

Reply:
Changed this to now create the instance immediately after validation.